R/fledgling.R

```diff
@@ -28,44 +28,132 @@ read_version <- function() {
-  # reinventing the wheel https://pandoc.org/MANUAL.html#options-affecting-specific-writers
-}
+read_news <- function(news_lines = NULL) {
+  if (is.null(news_lines)) {
+    if (file.exists("NEWS.md")) {
+      news_lines <- readLines("NEWS.md")
+    } else {
+      news_lines <- character()
+    }
+  }
 
-read_news <- function() {
-  if (file.exists("NEWS.md")) {
-    news <- readLines("NEWS.md")
-  } else {
-    news <- character()
+  news <- parse_news_md(news_lines, strict = TRUE)
+
+  if (is.null(news)) {
+    return(
+      list(
+        section_df = NULL,
+        preamble = news_preamble()
+      )
+    )
   }
 
-  parse_news(news)
-}
+  # NEWS content under no version
+  if (length(news) == 1 && !nzchar(names(news))) {
+    rlang::abort("NEWS content not under a version header")
+  }
 
-parse_news <- function(news) {
-  header_rx <- '^#(?<h2>#)? +[a-zA-Z ]*?[a-zA-Z][a-zA-Z0-9.]*[a-zA-Z0-9] +(?<version>v?[0-9][0-9.-]*) *(?<date>\\(.*\\))? *(?<nickname>".*")?$'
+  # match parsed headers to the Markdown
+  get_section_start <- function(section_title, news_lines) {
+    escaped_section_title <- sub("\\(", "\\\\(", section_title)
+    escaped_section_title <- sub("\\)", "\\\\)", escaped_section_title)
+    section_start <- which(grepl(sprintf("^#+ +%s[:space:]?$", escaped_section_title), news_lines))
 
-  first_level_headers <- grep(header_rx, news, perl = TRUE)
-  start <- first_level_headers
-  end <- c(first_level_headers[-1] - 1L, length(news))
+    if (length(section_start) == 0) {
+      # what to do
+    }
 
-  section_df <- rematch2::re_match(news[first_level_headers], header_rx)[1:4]
-  section_df <- tibble::add_column(section_df, line = first_level_headers, .before = 1)
-  section_df <- tibble::add_column(section_df, original = news[first_level_headers])
-  section_df$h2 <- (section_df$h2 == "#")
-  section_df$news <- map2(start + 1L, end, ~ trim_empty_lines(news[seq2(.x, .y)]))
-  section_df$raw <- map2_chr(start, end, ~ paste(news[seq2(.x, .y)], collapse = "\n"))
+    if (length(section_start) > 1) {
+      # what to do
+    }
+    section_start
+  }
+
+  starts <- purrr::map_int(names(news), get_section_start, news_lines)
 
-  if (length(first_level_headers) == 0) {
-    preamble <- news
-  } else if (first_level_headers[[1]] == 1) {
-    preamble <- NULL
+  ends <- if (length(starts) == 1) {
+    length(news_lines)
   } else {
-    preamble <- trim_empty_lines(news[seq2(1, first_level_headers[[1]] - 1)])
+    c(starts[seq_along(starts[-1]) + 1] - 1, length(news_lines))
   }
 
+  section_df <- tibble::tibble(
+    start = starts,
+    end = ends,
+    h2 = grepl("##", news_lines[starts]), # TODO does not account for all syntaxes,
+    raw = map2_chr(starts, ends, ~ paste(news_lines[seq2(.x, .y)], collapse = "\n")),
+    news = split(news, seq_len(length(news)))
+  )
+
+  section_df$section_state <- "keep"
+
+  section_df$title <- names(news)
+
+  find_version <- function(text) {
```

Comment:
probably this should go into a function that takes a title and parses into the different component. A function that'd be tested.

Reply:
see #610 :innocent: